### PR TITLE
feat(Overlay): new React Overlay component

### PR DESCRIPTION
**Pull request by marashad001**
#924 Overlay component

## Changes
- Created new React `Overlay` component from `Progress Indicator` html demo page.

## Testing

- Tests if the component is visible on the page and blurs the appropriate content wrapped within it.

## Docs

Added changeset `sharp-pianos-destroy.mdx` and updated `progress-indicator.mdx`.

### CODE REVIEW

File: packages/itwinui-react/src/core/ProgressIndicators/Overlay/Overlay.test.tsx
Comment (marashad001) [reply packages/itwinui-react/src/core/ProgressIndicators/Overlay/Overlay.test.tsx:1]: Done!

File: packages/itwinui-react/src/core/ProgressIndicators/Overlay/Overlay.tsx
Comment (marashad001) [reply packages/itwinui-react/src/core/ProgressIndicators/Overlay/Overlay.tsx:10]: Done!

File: packages/itwinui-react/src/core/ProgressIndicators/index.ts
Comment (gretanausedaite) [inline packages/itwinui-react/src/core/ProgressIndicators/index.ts:9]: Remove

File: packages/itwinui-react/src/core/index.ts
Comment (gretanausedaite) [inline packages/itwinui-react/src/core/index.ts:115]: Importing from wrong location

**Comment by mayank99** [comment]
I think this implementation will not be enough. There are two parts to the overlay:
1. the content underneath that gets blurred. this needs the `inert` attribute (check html code) to hide it from assistive technology and prevent focusing.
2. the content above the overlay which gets centered (e.g. progress indicators).

So it might make sense to use composition or an extra prop:
```jsx
<Overlay>
  <Overlay.Message>
    loading...
  </Overlay.Message>
  <Overlay.HiddenContent>
    everything else goes here...
  </Overlay.HiddenContent>
</Overlay>
```
(names debatable)

File: packages/itwinui-react/src/core/ProgressIndicators/index.ts
Comment (marashad001) [reply packages/itwinui-react/src/core/ProgressIndicators/index.ts:9]: done

File: packages/itwinui-react/src/core/index.ts
Comment (marashad001) [reply packages/itwinui-react/src/core/index.ts:115]: done

**Comment by mayank99** [comment]
Here's a suggestion for a more convenient api:
```jsx
<Overlay content={'loading...'}>
  content beneath the overlay...
</Overlay>
```

And if users want something more flexibility, they can alternatively use subcomponents:
```jsx
<Overlay.Wrapper>
  <Overlay.Overlay>
    loading...
  </Overlay.Overlay>
  <Overlay.HiddenContent>
    content beneath the overlay...
  </Overlay.HiddenContent>
</Overlay.Wrapper>
```

Both of these snippets should produce the same result. This avoids some of the boilerplate while keeping the flexibility.

File: apps/website/src/pages/docs/progressindicator.mdx
Comment (marashad001) [reply apps/website/src/pages/docs/progressindicator.mdx]: fixed

File: examples/Overlay.main.tsx
Comment (gretanausedaite) [inline examples/Overlay.main.tsx:9]: This example is empty. Update with new code.

File: .changeset/sharp-pianos-destroy.md
Comment (gretanausedaite) [inline .changeset/sharp-pianos-destroy.md:5]: List subcomponents

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (gretanausedaite) [inline packages/itwinui-react/src/core/Overlay/Overlay.tsx:17]: Add unit test for inert attribute.

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (gretanausedaite) [inline packages/itwinui-react/src/core/Overlay/Overlay.tsx:26]: Add unit test for classname.

**Comment by mayank99** [comment]
> So I was able to create the subcomponents for your suggestion, but I would need more information on using the condensed API with just `Overlay`. How would I be able to use both the generic `Overlay` and the `Overlay` with subcomponents while having both be functional?

So you'll be able to achieve this using four components instead of three. The fourth component will make use of the other three.

```jsx
const OverlayWrapper = /*...*/;
const OverlayHiddenContent = /*...*/;
const OverlayOverlay = /*...*/;

/* this one is new */
const OverlayComponent = React.forwardRef((props, forwardedRef) => {
  const { content, children, ...rest } = props;

  return (
    <OverlayWrapper {...rest}>
      <OverlayHiddenContent>{children}</OverlayHiddenContent>
      <OverlayOverlay>{content}</OverlayOverlay>
    </OverlayWrapper>
  );
});
```

and then your export will look something like:

```jsx
export const Overlay = Object.assign(OverlayComponent, {
  Wrapper: OverlayWrapper,
  HiddenContent: OverlayHiddenContent,
  Overlay: OverlayOverlay
});
```

**Comment by marashad001** [comment]
> > So I was able to create the subcomponents for your suggestion, but I would need more information on using the condensed API with just `Overlay`. How would I be able to use both the generic `Overlay` and the `Overlay` with subcomponents while having both be functional?
> 
> So you'll be able to achieve this using four components instead of three. The fourth component will make use of the other three.
> 
> ```js
> const OverlayWrapper = /*...*/;
> const OverlayHiddenContent = /*...*/;
> const OverlayOverlay = /*...*/;
> 
> /* this one is new */
> const OverlayComponent = React.forwardRef((props, forwardedRef) => {
>   const { content, children, ...rest } = props;
> 
>   return (
>     <OverlayWrapper {...rest}>
>       <OverlayHiddenContent>{children}</OverlayHiddenContent>
>       <OverlayOverlay>{content}</OverlayOverlay>
>     </OverlayWrapper>
>   );
> });
> ```
> 
> and then your export will look something like:
> 
> ```js
> export const Overlay = Object.assign(OverlayComponent, {
>   Wrapper: OverlayWrapper,
>   HiddenContent: OverlayHiddenContent,
>   Overlay: OverlayOverlay
> });
> ```

Gotcha, thanks.

File: .changeset/sharp-pianos-destroy.md
Comment (marashad001) [reply .changeset/sharp-pianos-destroy.md]: done

File: examples/Overlay.main.tsx
Comment (marashad001) [reply examples/Overlay.main.tsx]: done

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Overlay/Overlay.tsx:43]: this custom attribute has no effect. the resulting html needs to be `<div inert>`, not `<div custom-attribute='inert'>`

see docs: https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inert

Comment (mayank99) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx:43]: react doesn't fully support `inert` for some reason, so you'll need to use this hack:

```suggestion
    <Box {...{ inert: '' }} ref={ref} {...rest}>
```

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Overlay/Overlay.tsx:41]: `as: Box = 'div'` is incorrect. You need to import the `Box` component and use that

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Overlay/Overlay.tsx:53]: rename from Overlay.Message to Overlay.Overlay

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Overlay/Overlay.tsx:25]: remove eslint comments

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Overlay/Overlay.tsx:12]: doesn't need to be exported

File: packages/itwinui-css/src/progress-indicator/progress-indicator.scss
Comment (mayank99) [inline packages/itwinui-css/src/progress-indicator/progress-indicator.scss:5]: shouldn't be imported here.

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Overlay/Overlay.tsx:42]: we should add a lazy-loaded polyfill for older browsers where inert is not supported

see compatibility table, it's below 90%: https://caniuse.com/mdn-api_htmlelement_inert

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (marashad001) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx:53]: just for the display name or the subcomponent too?

Comment (marashad001) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx]: fixed

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (marashad001) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx]: fixed

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (marashad001) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx]: done

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (marashad001) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx]: fixed

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (gretanausedaite) [inline packages/itwinui-react/src/core/Overlay/Overlay.tsx:103]:
```suggestion
  const modulePath = 'https://unpkg.com/wicg-inert@3.1.2/dist/inert.min.js';
```

Let's switch to unpkg to make it same as progress-indicator.html

File: apps/website/src/pages/docs/progressindicator.mdx
Comment (gretanausedaite) [inline apps/website/src/pages/docs/progressindicator.mdx:42]: Add a little bit of text explaining that users can use progress indicators with overlay.

Comment (gretanausedaite) [reply apps/website/src/pages/docs/progressindicator.mdx:42]: Or remove at all as you added new mdx file.

File: apps/website/src/pages/docs/overlay.mdx
Comment (gretanausedaite) [inline apps/website/src/pages/docs/overlay.mdx:7]: Overlay is not in inputs. Move it to core.

File: apps/website/src/pages/docs/overlay.mdx
Comment (gretanausedaite) [inline apps/website/src/pages/docs/overlay.mdx:16]: remove `truncate={false}` from both examples.

File: apps/website/src/pages/docs/progressindicator.mdx
Comment (marashad001) [reply apps/website/src/pages/docs/progressindicator.mdx:42]: done :)

File: apps/website/src/pages/docs/overlay.mdx
Comment (marashad001) [reply apps/website/src/pages/docs/overlay.mdx:7]: done :)

Comment (marashad001) [reply apps/website/src/pages/docs/overlay.mdx:7]: done

File: apps/website/src/pages/docs/overlay.mdx
Comment (marashad001) [reply apps/website/src/pages/docs/overlay.mdx:16]: done

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (marashad001) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx:103]: Ok thanks for the heads up

File: examples/Overlay.main.tsx
Comment (mayank99) [inline examples/Overlay.main.tsx:13]: some of these styles (specifically `marginBottom` and `position`) are unnecessary

same comment in `Overlay.sub.tsx`

File: .changeset/sharp-pianos-destroy.md
Comment (mayank99) [inline .changeset/sharp-pianos-destroy.md:1]: add a separate changeset mentioning that `iui-progress-overlay` has been renamed to `iui-overlay` and moved to `overlay.scss`

File: apps/storybook/cypress-visual-screenshots/baseline/Overlay.test.ts-Radial.png
Comment (mayank99) [inline apps/storybook/cypress-visual-screenshots/baseline/Overlay.test.ts-Radial.png:1]: update your branch, then re-approve this test image. ProgressRadial was changed in #1356 so it will look different now.

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Overlay/Overlay.tsx:62]:
```suggestion
const OverlayWrapper = polymorphic('iui-overlay-wrapper');
```

File: examples/Overlay.main.tsx
Comment (marashad001) [reply examples/Overlay.main.tsx:13]: I removed `marginBottom`, `maxWidth`, and `maxHeight`. `position` is there to keep the Overlay from being applied over the entire LiveExample viewport (it was covering the codesandbox icon in the bottom left).

Comment (mayank99) [reply examples/Overlay.main.tsx:13]: but OverlayWrapper already has `position: relative` in the css

File: .changeset/sharp-pianos-destroy.md
Comment (marashad001) [reply .changeset/sharp-pianos-destroy.md:1]: Gotcha

File: examples/Overlay.main.tsx
Comment (marashad001) [reply examples/Overlay.main.tsx]: what could I put to keep the wrapper from overtaking the viewport?

File: apps/storybook/cypress-visual-screenshots/baseline/Overlay.test.ts-Radial.png
Comment (marashad001) [reply apps/storybook/cypress-visual-screenshots/baseline/Overlay.test.ts-Radial.png:1]: done

File: examples/Overlay.main.tsx
Comment (mayank99) [reply examples/Overlay.main.tsx]: nothing. OverlayWrapper already has `position: relative`, so it should look fine with just this code:

```jsx
import * as React from 'react';
import { Overlay } from '@itwin/itwinui-react';

export default () => {
  return (
    <Overlay
      content={'loading...'}
      style={{ border: '1px solid var(--iui-color-border)' }}
    >
      <div>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
        veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea
        commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
        velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint
        occaecat cupidatat non proident, sunt in culpa qui officia deserunt
        mollit anim id est laborum.
      </div>
    </Overlay>
  );
};

```

Comment (marashad001) [reply examples/Overlay.main.tsx]: image

this is what it looks like, the `Overlay` takes over the entire LiveExample. I was asking to find a way to confine it to just the sample text inside the `Overlay` component.

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (marashad001) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx:55]: ^^ the `Overlay` stopped working properly with the `polymorphic`.

File: examples/Overlay.main.tsx
Comment (mayank99) [reply examples/Overlay.main.tsx]: For some reason you've reverted the other change to `OverlayWrapper`. https://github.com/iTwin/iTwinUI/pull/1328#discussion_r1242385470

If you change it back, it works correctly.

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (mayank99) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx:55]: what do you mean "stopped working properly"? what problem are you facing? `children` prop works fine with it; see https://github.com/iTwin/iTwinUI/pull/1328#discussion_r1246915474

File: examples/Overlay.main.tsx
Comment (marashad001) [reply examples/Overlay.main.tsx]: Ok, got it working correctly, thanks

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (marashad001) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx:62]: got it working

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (marashad001) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx]: I got it working again thanks

File: .changeset/real-penguins-admire.md
Comment (mayank99) [inline .changeset/real-penguins-admire.md:2]:
```suggestion
'@itwin/itwinui-react': major
```

**Review by mayank99 — APPROVED** [review]

File: .changeset/real-penguins-admire.md
Comment (mayank99) [inline .changeset/real-penguins-admire.md:2]: just noticed it's the wrong package name

```suggestion
'@itwin/itwinui-css': major
```

**Comment by mayank99** [comment]
website build needs to be fixed. https://github.com/iTwin/iTwinUI/actions/runs/5423781577/jobs/9862225841?pr=1328

Also wait for approval from second reviewer (@adamhock) before merging.

File: packages/itwinui-react/tsconfig.json
Comment (adamhock) [inline packages/itwinui-react/tsconfig.json:5]: What is the reason for this change?

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (adamhock) [inline packages/itwinui-react/src/core/Overlay/Overlay.tsx:64]: Should change the `Overlay.Wrapper`

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (adamhock) [inline packages/itwinui-react/src/core/Overlay/Overlay.tsx:63]: Maybe add a second example which uses `Overlay` without any subcomponents

Comment (marashad001) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx:64]: fixed

File: packages/itwinui-react/src/core/Overlay/Overlay.tsx
Comment (marashad001) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx]: fixed

Comment (marashad001) [reply packages/itwinui-react/src/core/Overlay/Overlay.tsx]: thanks good catch

File: packages/itwinui-react/tsconfig.esm.json
Comment (marashad001) [reply packages/itwinui-react/tsconfig.esm.json]: reverted

File: packages/itwinui-react/tsconfig.json
Comment (marashad001) [reply packages/itwinui-react/tsconfig.json:5]: This is for the `useInertPolyfill` function inside `Overlay.tsx`.

**Pull request merged**
